File: core/parser.py
import os
import re
from typing import Tuple, List

import chardet
import pandas as pd

LANG_PATTERNS = {
    "KO": re.compile(r"^(KO|Korean|한국어|ko-KR|kor)$", re.IGNORECASE),
    "EN": re.compile(r"^(EN|English|영어|en-US|en-GB|eng)$", re.IGNORECASE),
    "JP": re.compile(r"^(JP|JA|Japanese|일본어|ja-JP|jpn)$", re.IGNORECASE),
    # Chinese Traditional (Taiwan) and Simplified (PRC) as separate targets
    "CT": re.compile(r"^(CT|CHT|zh-TW|Chinese[\s_\-]?Traditional|Chinese[\s_\-]?\(?\s*Taiwan\s*\)?|중국어[\s_\-]?번체|繁體中文)$", re.IGNORECASE),
    "CS": re.compile(r"^(CS|CHS|zh-CN|Chinese[\s_\-]?Simplified|Chinese[\s_\-]?\(?\s*PRC\s*\)?|중국어[\s_\-]?간체|简体中文)$", re.IGNORECASE),
    "ZH": re.compile(r"^(ZH|Chinese|중국어|zho)$", re.IGNORECASE),
    "DE": re.compile(r"^(DE|German|독일어|de-DE|deu)$", re.IGNORECASE),
    "FR": re.compile(r"^(FR|French|프랑스어|fr-FR|fra)$", re.IGNORECASE),
    "ES": re.compile(r"^(ES|Spanish|스페인어|es-ES|spa)$", re.IGNORECASE),
    "PT": re.compile(r"^(PT|Portuguese|Portuguese[\s_\-]?\(?\s*Brazil\s*\)?|포르투갈어|pt-BR|pt-PT|por)$", re.IGNORECASE),
    "IT": re.compile(r"^(IT|Italian|이탈리아어|it-IT|ita)$", re.IGNORECASE),
    "RU": re.compile(r"^(RU|Russian|러시아어|ru-RU|rus)$", re.IGNORECASE),
    "TH": re.compile(r"^(TH|Thai|태국어|th-TH|tha)$", re.IGNORECASE),
    "VI": re.compile(r"^(VI|Vietnamese|베트남어|vi-VN|vie)$", re.IGNORECASE),
    "ID": re.compile(r"^(ID|Indonesian|인도네시아어|id-ID|ind)$", re.IGNORECASE),
    "AR": re.compile(r"^(AR|Arabic|아랍어|ar-SA|ara)$", re.IGNORECASE),
    "TR": re.compile(r"^(TR|Turkish|터키어|tr-TR|tur)$", re.IGNORECASE),
}

# Columns that contain relevant supplementary info (shown after translations)
_INFO_COL_RE = re.compile(
    r"^(note|notes|분류|category|comment|comments|remark|remarks|memo|description|설명)$",
    re.IGNORECASE,
)

# All recognized language codes (for classification)
_LANG_CODES = set(LANG_PATTERNS.keys())

# Common prefixes/suffixes stripped before matching language patterns
_COL_PREFIXES = re.compile(
    r"^(target|tgt|source|src|lang|translation|trans)[_\-\s.:]+",
    re.IGNORECASE,
)
_COL_SUFFIXES = re.compile(
    r"[_\-\s.:](target|tgt|source|src|lang|translation|trans)$",
    re.IGNORECASE,
)

HANGUL_RE = re.compile(r"[\uAC00-\uD7AF]")
MAX_ENTRIES = 500_000


def normalize_column_name(col_name: str) -> str:
    """Normalize a column name to a standard language code.

    Examples: "Target_EN" → "EN", "English" → "EN", "영어" → "EN",
              "ja-JP" → "JP", "tgt_French" → "FR".
    Returns the original name unchanged if no language pattern matches.
    """
    col_stripped = str(col_name).strip()

    # Try direct match first
    for lang_code, pattern in LANG_PATTERNS.items():
        if pattern.match(col_stripped):
            return lang_code

    # Strip common prefixes/suffixes and retry
    cleaned = _COL_PREFIXES.sub("", col_stripped)
    cleaned = _COL_SUFFIXES.sub("", cleaned).strip()
    if cleaned and cleaned != col_stripped:
        for lang_code, pattern in LANG_PATTERNS.items():
            if pattern.match(cleaned):
                return lang_code

    return col_stripped


def classify_column(norm_name: str) -> int:
    """Classify a normalized column name for display ordering.

    Returns:
        0 — translation column (known language code)
        1 — relevant info column (Note, 분류, etc.)
        2 — metadata / non-translation column (pushed to back)
    """
    if norm_name in _LANG_CODES:
        return 0
    if _INFO_COL_RE.match(norm_name):
        return 1
    return 2


def detect_encoding(file_path: str) -> str:
    with open(file_path, "rb") as f:
        raw = f.read(100_000)
    result = chardet.detect(raw)
    encoding = result.get("encoding", "utf-8") or "utf-8"
    enc_lower = encoding.lower().replace("-", "")
    if enc_lower in ("ascii", "utf8"):
        return "utf-8"
    return encoding


def _detect_delimiter(file_path: str, encoding: str) -> str:
    with open(file_path, "r", encoding=encoding, errors="replace") as f:
        sample = f.read(8192)
    counts = {"\t": sample.count("\t"), ",": sample.count(","), ";": sample.count(";")}
    if counts["\t"] > 0:
        return "\t"
    return max(counts, key=counts.get)


def parse_file(file_path: str) -> pd.DataFrame:
    ext = os.path.splitext(file_path)[1].lower()

    if ext == ".xlsx":
        try:
            df = pd.read_excel(file_path, sheet_name=0, dtype=str)
        except Exception as e:
            raise ValueError(f"Could not read {os.path.basename(file_path)}: {e}")
        xls = pd.ExcelFile(file_path)
        df.attrs["extra_sheets"] = len(xls.sheet_names) - 1
        df.attrs["sheet_names"] = xls.sheet_names

    elif ext in (".csv", ".txt"):
        encoding = detect_encoding(file_path)
        delimiter = _detect_delimiter(file_path, encoding)
        # For .txt files, assume no header row
        header_arg = None if ext == ".txt" else "infer"
        try:
            df = pd.read_csv(file_path, sep=delimiter, encoding=encoding, dtype=str,
                             on_bad_lines="skip", header=header_arg)
        except pd.errors.EmptyDataError:
            raise ValueError(f"No entries found in {os.path.basename(file_path)}")
        except Exception:
            for fallback_enc in ["utf-8", "euc-kr", "cp949", "latin-1"]:
                try:
                    df = pd.read_csv(file_path, sep=delimiter, encoding=fallback_enc, dtype=str,
                                     on_bad_lines="skip", header=header_arg)
                    break
                except pd.errors.EmptyDataError:
                    raise ValueError(f"No entries found in {os.path.basename(file_path)}")
                except Exception:
                    continue
            else:
                raise ValueError(f"Could not read {os.path.basename(file_path)}")
    else:
        raise ValueError(f"Unsupported format: {ext}. Use .xlsx, .csv, or .txt")

    df = df.dropna(how="all").reset_index(drop=True)
    if len(df) == 0:
        raise ValueError(f"No entries found in {os.path.basename(file_path)}")
    df.attrs["source_file"] = os.path.basename(file_path)
    return df


def detect_columns(df: pd.DataFrame) -> Tuple[str, List[str]]:
    columns = list(df.columns)
    source = None

    for col in columns:
        col_stripped = str(col).strip()
        for lang_code, pattern in LANG_PATTERNS.items():
            if pattern.match(col_stripped):
                if lang_code == "KO":
                    source = col
                break

    if source is not None:
        targets = [c for c in columns if c != source]
        return source, targets

    scan_rows = min(10, len(df))
    korean_counts = {}
    for col in columns:
        count = 0
        for i in range(scan_rows):
            val = str(df.iloc[i][col]) if pd.notna(df.iloc[i][col]) else ""
            if HANGUL_RE.search(val):
                count += 1
        if count > 0:
            korean_counts[col] = count

    if korean_counts:
        for col in columns:
            if col in korean_counts:
                source = col
                break
        targets = [c for c in columns if c != source]
        return source, targets

    source = columns[0]
    targets = columns[1:] if len(columns) > 1 else []
    return source, targets


def check_entry_limit(current_total: int, new_count: int, file_name: str) -> None:
    if current_total + new_count > MAX_ENTRIES:
        raise ValueError(
            f"Cannot load {file_name}: total entries would exceed 500K limit "
            f"({current_total} + {new_count} = {current_total + new_count})"
        )

File: web/src/core/parser.ts
/**
 * Core parser — TypeScript port of desktop `core/parser.py`.
 *
 * Handles file parsing (xlsx / csv / txt), column normalisation,
 * language detection and entry-limit checking.
 */

import * as XLSX from "xlsx";
import Papa from "papaparse";
import jschardet from "jschardet";

/* ------------------------------------------------------------------ */
/*  Language patterns                                                   */
/* ------------------------------------------------------------------ */

export const LANG_PATTERNS: Record<string, RegExp> = {
  KO: /^(KO|Korean|한국어|ko-KR|kor|Source)$/i,
  EN: /^(EN|English|영어|en-US|en-GB|eng)$/i,
  JP: /^(JP|JA|Japanese|일본어|ja-JP|jpn)$/i,
  CT: /^(CT|CHT|zh-TW|zh-Hant|TW|Chinese[\s_-]?Traditional|Chinese[\s_-]?\(?\s*Taiwan\s*\)?|중국어[\s_-]?번체|繁體中文)$/i,
  CS: /^(CS|CHS|zh-CN|zh-Hans|Chinese[\s_-]?Simplified|Chinese[\s_-]?\(?\s*PRC\s*\)?|중국어[\s_-]?간체|简体中文)$/i,
  ZH: /^(ZH|Chinese|중국어|zho)$/i,
  DE: /^(DE|German|독일어|de-DE|deu)$/i,
  FR: /^(FR|French|프랑스어|fr-FR|fra)$/i,
  ES: /^(ES|Spanish|스페인어|es-ES|spa|ES-LATAM|Spanish[\s_-]?\(?\s*Latin\s*America\s*\)?)$/i,
  PT: /^(PT|Portuguese|Portuguese[\s_-]?\(?\s*Brazil\s*\)?|포르투갈어|pt-BR|pt-PT|por)$/i,
  IT: /^(IT|Italian|이탈리아어|it-IT|ita)$/i,
  RU: /^(RU|Russian|러시아어|ru-RU|rus)$/i,
  TH: /^(TH|Thai|태국어|th-TH|tha)$/i,
  VI: /^(VI|Vietnamese|베트남어|vi-VN|vie)$/i,
  ID: /^(ID|Indonesian|인도네시아어|id-ID|ind)$/i,
  AR: /^(AR|Arabic|아랍어|ar-SA|ara)$/i,
  TR: /^(TR|Turkish|터키어|tr-TR|tur)$/i,
};

const _INFO_COL_RE =
  /^(note|notes|분류|category|comment|comments|remark|remarks|memo|description|설명)$/i;

const _LANG_CODES = new Set(Object.keys(LANG_PATTERNS));

const _COL_PREFIXES =
  /^(target|tgt|source|src|lang|translation|trans)[_\-\s.:]+/i;
const _COL_SUFFIXES =
  /[_\-\s.:](target|tgt|source|src|lang|translation|trans)$/i;

const HANGUL_RE = /[\uAC00-\uD7AF]/;

const MAX_ENTRIES = 500_000;

/* ------------------------------------------------------------------ */
/*  Column helpers                                                     */
/* ------------------------------------------------------------------ */

/**
 * Normalize a column header to a standard language code.
 *
 * "Target_EN" → "EN", "English" → "EN", "영어" → "EN",
 * "ja-JP" → "JP", "tgt_French" → "FR".
 *
 * Returns the original (trimmed) name unchanged if no pattern matches.
 */
export function normalizeColumnName(colName: string): string {
  const colStripped = String(colName).trim();

  // Direct match
  for (const [langCode, pattern] of Object.entries(LANG_PATTERNS)) {
    if (pattern.test(colStripped)) return langCode;
  }

  // Strip common prefixes / suffixes and retry
  let cleaned = colStripped.replace(_COL_PREFIXES, "");
  cleaned = cleaned.replace(_COL_SUFFIXES, "").trim();

  if (cleaned && cleaned !== colStripped) {
    for (const [langCode, pattern] of Object.entries(LANG_PATTERNS)) {
      if (pattern.test(cleaned)) return langCode;
    }
  }

  return colStripped;
}

/**
 * Classify a normalized column name for display ordering.
 *
 * 0 — language column (known code)
 * 1 — info column (note, category, comment …)
 * 2 — metadata / non-translation column
 */
export function classifyColumn(normName: string): 0 | 1 | 2 {
  if (_LANG_CODES.has(normName)) return 0;
  if (_INFO_COL_RE.test(normName)) return 1;
  return 2;
}

/* ------------------------------------------------------------------ */
/*  Encoding detection                                                 */
/* ------------------------------------------------------------------ */

function detectEncoding(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer).slice(0, 100_000);
  const result = jschardet.detect(
    // jschardet wants a binary string
    Array.from(bytes)
      .map((b) => String.fromCharCode(b))
      .join(""),
  );
  const encoding = result.encoding ?? "utf-8";
  const encLower = encoding.toLowerCase().replace(/-/g, "");
  if (encLower === "ascii" || encLower === "utf8") return "utf-8";
  return encoding;
}

function decodeBuffer(buffer: ArrayBuffer, encoding: string): string {
  try {
    const decoder = new TextDecoder(encoding);
    return decoder.decode(buffer);
  } catch {
    // Fallback chain
    for (const fb of ["utf-8", "euc-kr", "iso-8859-1"]) {
      try {
        return new TextDecoder(fb).decode(buffer);
      } catch {
        continue;
      }
    }
    // Last resort — latin1 never throws
    return new TextDecoder("iso-8859-1").decode(buffer);
  }
}

/* ------------------------------------------------------------------ */
/*  File parsing                                                       */
/* ------------------------------------------------------------------ */

export interface ParseResult {
  /** Array of row objects (column → value). */
  data: Record<string, string>[];
  /** Ordered column names. */
  columns: string[];
  /** Number of additional sheets (xlsx only, else 0). */
  extraSheets: number;
  /** Original file name. */
  sourceFile: string;
}

/**
 * Parse an uploaded File (xlsx / csv / txt) and return structured data.
 */
export async function parseFile(file: File): Promise<ParseResult> {
  const ext = file.name.slice(file.name.lastIndexOf(".")).toLowerCase();
  const buffer = await file.arrayBuffer();

  if (ext === ".xlsx") {
    return parseXlsx(buffer, file.name);
  } else if (ext === ".csv" || ext === ".txt") {
    return parseCsvTxt(buffer, ext, file.name);
  }

  throw new Error(
    `Unsupported format: ${ext}. Use .xlsx, .csv, or .txt`,
  );
}

function parseXlsx(buffer: ArrayBuffer, fileName: string): ParseResult {
  let workbook: XLSX.WorkBook;
  try {
    workbook = XLSX.read(buffer, { type: "array" });
  } catch (e) {
    throw new Error(
      `Could not read ${fileName}: ${e instanceof Error ? e.message : e}`,
    );
  }

  const sheetName = workbook.SheetNames[0];
  const sheet = workbook.Sheets[sheetName];
  const rows: Record<string, string>[] = XLSX.utils.sheet_to_json(sheet, {
    defval: "",
    raw: false,
  });

  if (rows.length === 0) {
    throw new Error(`No entries found in ${fileName}`);
  }

  const columns = Object.keys(rows[0]);

  // Drop all-empty rows
  const data = rows.filter((row) =>
    columns.some((c) => row[c] !== undefined && String(row[c]).trim() !== ""),
  );

  if (data.length === 0) {
    throw new Error(`No entries found in ${fileName}`);
  }

  return {
    data,
    columns,
    extraSheets: workbook.SheetNames.length - 1,
    sourceFile: fileName,
  };
}

function parseCsvTxt(
  buffer: ArrayBuffer,
  ext: string,
  fileName: string,
): ParseResult {
  const encoding = detectEncoding(buffer);
  const text = decodeBuffer(buffer, encoding);

  const hasHeader = ext !== ".txt";

  const parsed = Papa.parse<Record<string, string>>(text, {
    header: hasHeader,
    delimiter: "", // auto-detect
    skipEmptyLines: "greedy",
    dynamicTyping: false,
  });

  if (parsed.data.length === 0) {
    throw new Error(`No entries found in ${fileName}`);
  }

  let columns: string[];
  let data: Record<string, string>[];

  if (hasHeader) {
    columns = parsed.meta.fields ?? Object.keys(parsed.data[0]);
    data = parsed.data;
  } else {
    // txt — no header; generate Col0, Col1, …
    const firstRow = parsed.data[0];
    const nCols = Object.keys(firstRow).length;
    columns = Array.from({ length: nCols }, (_, i) => `Col${i}`);

    // Papa with header:false gives arrays disguised as objects keyed "0","1",…
    data = (parsed.data as unknown as string[][]).map((row) => {
      const obj: Record<string, string> = {};
      columns.forEach((c, i) => {
        obj[c] = row[i] ?? "";
      });
      return obj;
    });
  }

  // Drop all-empty rows
  data = data.filter((row) =>
    columns.some((c) => row[c] !== undefined && String(row[c]).trim() !== ""),
  );

  if (data.length === 0) {
    throw new Error(`No entries found in ${fileName}`);
  }

  return { data, columns, extraSheets: 0, sourceFile: fileName };
}

/* ------------------------------------------------------------------ */
/*  Column detection                                                   */
/* ------------------------------------------------------------------ */

export interface ColumnDetection {
  source: string;
  targets: string[];
}

/**
 * Detect source (Korean) column and target columns.
 *
 * 1. Check language patterns — if KO matches, that's the source.
 * 2. Scan first 10 rows for Hangul content.
 * 3. Fall back to first column.
 */
export function detectColumns(
  data: Record<string, string>[],
  columns: string[],
): ColumnDetection {
  let source: string | null = null;

  // 1. Header-based detection
  for (const col of columns) {
    const colStripped = String(col).trim();
    for (const [langCode, pattern] of Object.entries(LANG_PATTERNS)) {
      if (pattern.test(colStripped)) {
        if (langCode === "KO") {
          source = col;
        }
        break;
      }
    }
  }

  if (source !== null) {
    return { source, targets: columns.filter((c) => c !== source) };
  }

  // 2. Content scan for Hangul
  const scanRows = Math.min(10, data.length);
  const koreanCounts: Record<string, number> = {};

  for (const col of columns) {
    let count = 0;
    for (let i = 0; i < scanRows; i++) {
      const val = data[i]?.[col];
      if (val != null && HANGUL_RE.test(String(val))) {
        count++;
      }
    }
    if (count > 0) {
      koreanCounts[col] = count;
    }
  }

  if (Object.keys(koreanCounts).length > 0) {
    // Pick the first column (in original order) that has Korean
    for (const col of columns) {
      if (col in koreanCounts) {
        source = col;
        break;
      }
    }
    return { source: source!, targets: columns.filter((c) => c !== source) };
  }

  // 3. Fallback — first column
  source = columns[0];
  return {
    source,
    targets: columns.length > 1 ? columns.slice(1) : [],
  };
}

/* ------------------------------------------------------------------ */
/*  Entry limit                                                        */
/* ------------------------------------------------------------------ */

/**
 * Throw if adding `newCount` entries to `currentTotal` would exceed 500K.
 */
export function checkEntryLimit(
  currentTotal: number,
  newCount: number,
  fileName: string,
): void {
  if (currentTotal + newCount > MAX_ENTRIES) {
    throw new Error(
      `Cannot load ${fileName}: total entries would exceed 500K limit ` +
        `(${currentTotal} + ${newCount} = ${currentTotal + newCount})`,
    );
  }
}
